Add NTP agent: __init__, config and ntp.py, update Vagrantfile.ntp

bodiakof opened the pull request.

bodiakof committed:
Add NTP agent: __init__, config and ntp.py, update Vagrantfile.ntp

bodiakof committed:
Add NTP agent: __init__, config and ntp.py

bodiakof commented:
Added modified .pre-commit-config.yaml file to this PR my mistake.

tsytsyksvitlana commented on agents/ntp/ntp.py:
```diff
@@ -0,0 +1,35 @@
+
+    def _is_port_open(self):
+        """Check UDP 123 instead of TCP."""
+        if self.port == -1:
```
duplicated logic: you are copying _is_port_open from ```ServerAgent``` . This method will be inherited
upd: ok, it seems like you changed udp to tcp, but I'm not sure which approach we should use

YelyzavetaV replied:
1. Agree with @tsytsyksvitlana: this shouldn't be duplicated in children agents.
2. If you want to change base agent's logic, it needs to be justified and discussed.

tsytsyksvitlana commented on vagrant/Vagrantfile.ntp:
```diff
@@ -1,9 +1,29 @@
 
+  config.vm.network "public_network",
+    ip:      ENV_NTP["IP"],
+    netmask: ENV_NTP["NETMASK"]
```
wait for @vladikD PR to be merged and use his version of ```Vagrantfile.ntp```

mehalyna commented on agents/ntp/ntp.py:
```diff
@@ -0,0 +1,35 @@
+        s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+        try:
+            s.settimeout(2)
+            s.sendto(b'', (self.ip, self.port))
```
Currently, you only send an empty UDP packet and assume success on no error. It would be better to validate a response (even partial) to confirm NTP is functioning, not just listening:
```
try:
    s.sendto(b'\x1b' + 47 * b'\0', (self.ip, self.port)) 
    data, _ = s.recvfrom(512)
    return bool(data)  # Or validate packet length
except socket.timeout:
    return False
```
This ensures that you're not just checking port availability but actual NTP behavior.

mehalyna commented on agents/ntp/ntp.py:
```diff
@@ -0,0 +1,35 @@
+            s.settimeout(2)
+            s.sendto(b'', (self.ip, self.port))
+            return True
+        except socket.error:
```
Add logging when `_is_port_open` fails:
```
except socket.error as e:
    self.logger.warning("NTP UDP check failed: %s", e)
    return False
```

mehalyna commented on vagrant/Vagrantfile.ntp:
```diff
@@ -1,9 +1,29 @@
+    # Install NTP
+    sudo apt update
+    sudo apt install -y ntp
+
```
You're installing ntp, but not enabling it. Add:
```
sudo systemctl enable ntp
sudo systemctl restart ntp
```
To ensure it starts on boot and applies the config.

mehalyna commented on vagrant/Vagrantfile.ntp:
```diff
@@ -1,9 +1,29 @@
+
+    # Install NTP
+    sudo apt update
+    sudo apt install -y ntp
```
To avoid silent failures, you could add:
```
if ! systemctl status ntp; then
  echo "NTP installation or service failed"
  exit 1
fi
```
Or log the status for debug:
```
sudo systemctl status ntp || true
```

YelyzavetaV commented on .pre-commit-config.yaml:
```diff
@@ -12,7 +12,7 @@ repos:
 
   - repo: https://github.com/PyCQA/flake8
-    rev: 7.1.1
+    rev: 5.0.4
```
Please suppress this from this PR: as you have already created a separate dedicated issue, it's better we fully isolate this.

bodiakof committed:
Modified Vagrantfile.ntp according to suggestions, removed .pre-commit-config.yaml from PR

bodiakof committed:
Modified _is_port_open method in agent.py so it's compatible with NTP server, removed _is_port_open from ntp.py

bodiakof committed:
Merge branch 'develop' into 19-create-and-test-agent-script-for-ntp-server

bodiakof committed:
Added .pre-commit-config.yaml file to pass GitHub Actions

bodiakof committed:
Changed flake8's version back to 7.1.1 (in .pre-commit-config.yaml) as it supposed to be :)

bodiakof committed:
Modified agents/ntp/config.ini and agents/ntp/ntp.py

YelyzavetaV committed:
feat: integrate protocol attribute into agent infrastracture

YelyzavetaV reviewed (changes requested):
Based on your code, I propose how the additional attribute `protocol` can be integrated consistently into the current agent infrastructure. Key points to consider:

- The original code relied on repetitive logic, I tried to minimize it (see `is_port_open`)
- We no longer pass default values as class attributes. What's done instead:
    - `PROTOCOLS` constant is defined to set the allowed values for `protocol`
    - The type and value checks for `protocol` are implemented via properties. Properties also handle the default value
    - Probe payload and packet sizes are not useful as attributes. Instead, they are implemented as function parameters for `is_port_open`
- Another important proposal is to make parent's `is_service_healthy` into an abstract method and override it in each child. The motivations are the following:
    - Each server should in fact implement server-specific health check. Your NTP agent currently contains the most complete check - the rest of the existing agents should be extended.
    - Such an approach allows us to pass server-specific parameter as default function arguments in `is_service_healthy`. For example, your binary string and `packet_size=48` go there.


What I'm asking you to do:

- Review what I propose. Let me know what you agree with and what you want to be done differently.
- Merge `develop` and make sure all children agents are consistent with the changes (this includes Web agent, which is currently not present in your branch).
- Make sure agents' tests are passing.

bodiakof committed:
Merge branch 'develop' into 19-create-and-test-agent-script-for-ntp-server

bodiakof committed:
Merge branch 'develop' into 19-create-and-test-agent-script-for-ntp-server

bodiakof committed:
Modified 'get_agent_status.py' (fixed 'no attribute logger' issue, supplemented with code from Server Agents at Wiki)

bodiakof committed:
Merge branch '19-create-and-test-agent-script-for-ntp-server' of https://github.com/ITA-Dnipro/PyDataCenter5000 into 19-create-and-test-agent-script-for-ntp-server

bodiakof committed:
Modified 'get_agent_status.py' (fixed 'no attribute logger' issue, supplemented with code from Server Agents at Wiki)

bodiakof committed:
Add post_data to ServerAgent and make MockAgent instantiable

- Implement post_data(...) with JSON payload, retries on URLError, and appropriate logging:
  • Logs “Retrying in X seconds…” on failures
  • Logs “Success on attempt N” and “POST request status: <code>” on success
  • Logs “All N attempts failed. Data not sent.” then raises RuntimeError after final retry
- In ServerAgent.__init__(), initialize self.log_path="" so MockAgent.__del__ won’t crash
- In setup_logging(), assign self.log_path to the configured log file
- Override service_healthy() in MockAgent to simply return True (removes abstract‐method error)

YelyzavetaV reviewed (changes requested):
Everything looks good, but web agent's tests are failing. What remains to be done: in `agents/web/web.py`, add `service_healthy` method (an override for an abstract parent method) - just like it's done in `DNSAgent` and `NTPAgent.`

bodiakof committed:
Merge branch 'develop' into 19-create-and-test-agent-script-for-ntp-server

bodiakof committed:
Modified 'agent.py' to resolve merge conflicts, updated service_helthy() in 'dns.py', 'ntp.py' and 'web.py'

YelyzavetaV committed:
fix: fix log_path bug in tests

YelyzavetaV reviewed (approved):
Good work.
